bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/metrics/MemoryMetricsSource.kt

```diff
@@ -0,0 +1,150 @@
+package com.bugsnag.android.performance.internal.metrics
+
+import android.app.ActivityManager
+import android.content.Context
+import android.os.Process
+import android.util.AndroidException
+import com.bugsnag.android.performance.internal.BugsnagClock
+import com.bugsnag.android.performance.internal.getActivityManager
+import com.bugsnag.android.performance.internal.util.FixedRingBuffer
+
+internal class MemoryMetricsSource(
+    private val appContext: Context,
+    samplingDelayMs: Long,
+    maxSampleCount: Int = DEFAULT_SAMPLE_COUNT,
+) : AbstractSampledMetricsSource<MemoryMetricsSnapshot>(samplingDelayMs), Runnable {
+    private val buffer = FixedRingBuffer(maxSampleCount) { MemorySampleData() }
+
+    private val runtime = Runtime.getRuntime()
+    private val activityManager = appContext.getActivityManager()
+
+    private val deviceMemory by lazy {
+        calculateTotalMemory()
+    }
+
+    override fun captureSample() {
+        val timestamp = BugsnagClock.currentUnixNanoTime()
+        buffer.put { sample ->
+            sample.freeMemory = runtime.freeMemory()
+            sample.totalMemory = runtime.totalMemory()
+            sample.timestamp = timestamp
+
+            val memoryInfo = activityManager?.getProcessMemoryInfo(intArrayOf(Process.myPid()))
+            if (memoryInfo?.size == 1) {
+                val memoryInfoSample = memoryInfo[0]
```

Comment:
For safety sake, we should have a few more null-checks in here. `memoryInfo` could be `null`, and we should double check that `memoryInfo[0]` is also not `null`.

```suggestion
            val memoryInfoSample = memoryInfo?.getOrNull(0)?.takeIf { memoryInfo?.size == 1 }
```